pkg/apphealth/health.go

```diff
@@ -0,0 +1,225 @@
+	AppStatusHealthy   uint8 = 1
+)
+
+var log = logger.NewLogger("dapr.apphealth")
+
+// AppHealth manages the health checks for the app.
+type AppHealth struct {
+	config       *Config
+	probeFn      ProbeFunction
+	changeCb     ChangeCallback
+	report       chan uint8
+	failureCount *atomic.Int32
+	lastReport   *atomic.Int64
+	queue        chan struct{}
+
+	// Minimum interval between reports, in microseconds.
+	// Used by ratelimitReports (as a variable so it can be changed in tests).
+	reportMinInterval int64
+}
+
+// ProbeFunction is the signature of the function that performs health probes.
+// Health probe functions return errors only in case of internal errors.
+// Network errors are considered probe failures, and should return nil as errors.
+type ProbeFunction func(context.Context) (bool, error)
+
+// ChangeCallback is the signature of the callback that is invoked when the app's health status changes.
+type ChangeCallback func(status uint8)
+
+// NewAppHealth creates a new AppHealth object.
+func NewAppHealth(config *Config, probeFn ProbeFunction) *AppHealth {
+	return &AppHealth{
+		config:            config,
+		probeFn:           probeFn,
+		report:            make(chan uint8, 1),
+		failureCount:      atomic.NewInt32(config.Threshold), // Initial state is unhealthy until we validate it
+		lastReport:        atomic.NewInt64(0),
+		reportMinInterval: 1e6,
+		queue:             make(chan struct{}, 1),
+	}
+}
+
+// OnHealthChange sets the callback that is invoked when the health of the app changes (app becomes either healthy or unhealthy).
+func (h *AppHealth) OnHealthChange(cb ChangeCallback) {
+	h.changeCb = cb
+}
+
+// StartProbes starts polling the app on the interval.
+func (h *AppHealth) StartProbes(ctx context.Context) {
+	if h.probeFn == nil {
+		log.Fatal("Cannot start probes with nil probe function")
+	}
+	if h.config.ProbeTimeout > h.config.ProbeInterval {
+		log.Fatal("App health checks probe timeouts must be smaller than probe intervals")
+	}
+
+	log.Info("App health probes starting")
+
+	go func() {
+		ticker := time.NewTicker(h.config.ProbeInterval)
+
+		for {
+			select {
+			case <-ctx.Done():
+				ticker.Stop()
+				log.Info("App health probes stopping")
+				return
+			case <-ticker.C:
+				log.Debug("Probing app health")
+				h.Enqueue()
+			case status := <-h.report:
+				log.Debug("Received health status report")
+				ticker.Reset(h.config.ProbeInterval)
+				h.setResult(status == AppStatusHealthy)
+			case <-h.queue:
+				// Run synchronously so the loop is blocked
+				h.doProbe(ctx)
+			}
+		}
+	}()
+}
+
+// Enqueue adds a new probe request to the queue
+func (h *AppHealth) Enqueue() {
+	// The queue has a capacity of 1, so no more than one iteration can be queued up
+	select {
+	case h.queue <- struct{}{}:
+		// Do nothing
+	default:
+		// Do nothing
+	}
+}
+
+// ReportHealth is used by the runtime to report a health signal from the app.
+func (h *AppHealth) ReportHealth(status uint8) {
```

Comment:
This method is currently unused. It was implemented to support using other signals to report health, but the downstream implementation is not (yet) done. Nevertheless, I don't think there's harm in leaving this here, as it also comes with unit tests